enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj

```diff
@@ -264,45 +263,33 @@
   (mbql.u/replace m
-    (_ :guard (every-pred map? (complement :gtap?) :source-table #(get table-id->gtap (:source-table %))))
+    (_ :guard (every-pred map? (complement ::gtap?) :source-table #(get table-id->gtap (:source-table %))))
     (let [updated             (apply-gtap &match (get table-id->gtap (:source-table &match)))
           ;; now recursively apply gtaps anywhere else they might exist at this level, e.g. `:joins`
           recursively-updated (merge
                                (select-keys updated [:source-table :source-query])
                                (apply-gtaps (dissoc updated :source-table :source-query) table-id->gtap))]
-      ;; add a `:gtap?` key next to every `:source-table` key so when we do a second pass after adding JOINs they
+      ;; add a `::gtap?` key next to every `:source-table` key so when we do a second pass after adding JOINs they
       ;; don't get processed again
       (mbql.u/replace recursively-updated
         (_ :guard (every-pred map? :source-table))
-        (assoc &match :gtap? true)))))
-
-(defn- id->col-info [query field-id]
-  (when field-id
-    (annotate/col-info-for-field-clause (:query query) [:field-id field-id])))
-
-(defn- update-col-metadata [query field-name->id-delay {:keys [id source], field-ref :field_ref, field-name :name, :as col}]
-  (let [id (or id (when (and (= (first field-ref) :field-literal)
-                             field-name)
-                    (get @field-name->id-delay field-name)))]
-    (merge
-     col
-     (id->col-info query id)
-     (when (= source :native)
-       {:source :fields}))))
+        (assoc &match ::gtap? true)))))
 
 (defn- merge-metadata
-  "Merge column metadata from the source Table into the current results `metadata`. This way the final results metadata
-  coming back matches what we'd get if the query was not running with a GTAP."
-  [query metadata]
-  (let [source-table-id      (mbql.u/query->source-table-id query)
-        field-name->id-delay (delay
-                               (u/prog1 (when source-table-id
-                                          (db/select-field->id :name Field :table_id source-table-id))
-                                 (qp.store/fetch-and-store-fields! (vals <>))))]
-    (update metadata :cols (fn [cols]
-                             (mapv (partial update-col-metadata query field-name->id-delay) cols)))))
+  "Merge column metadata from the non-GTAPped version of the query into the GTAPped results `metadata`. This way the
+  final results metadata coming back matches what we'd get if the query was not running with a GTAP."
+  [original-query metadata]
+  (letfn [(merge-cols [cols]
+            (let [expected-cols          (binding [*current-user-permissions-set* (atom #{"/"})]
+                                           ((requiring-resolve 'metabase.query-processor/query->expected-cols) original-query))
```

Comment:
`query->expected-cols` throws if the query-type is not `:query`. Is that constraint satisfied by all pathways here?

Reply:
original query is always going to be MBQL, we can't apply a GTAP (sandbox) to a native query